maro/rl/policy/policy_manager.py

```diff
@@ -339,3 +339,193 @@ def exit(self):
+                    self._trainer2policies[f"TRAINER.{trainer_id}"].append(name)
+            else:
+                trainer_id_list = list(range(self.num_trainers))
+                for i, name in enumerate(self.policy_dict):
+                    for trainer_id in trainer_id_list[i::len(self.policy_dict)]:
+                        self._policy2trainer[name].append(f"TRAINER.{trainer_id}")
+                        self._trainer2policies[f"TRAINER.{trainer_id}"].append(name)
+
+        # allocate trainers according to historical experience numbers.
+        else:
+            total_num_policy = sum(self._num_experiences_by_policy.values())
+            average_payload = total_num_policy / self.num_trainers
+
+            offset = 0
+            policy_quota = dict()
+            for name, num_exp in self._num_experiences_by_policy.items():
+                quota = num_exp / average_payload
+                quota = max(1, int(round(quota)))
+                policy_quota[name] = quota
+
+            # adjust quota if any redundancy occurs.
+            redundancy = self.num_trainers - sum(policy_quota.values())
+            if redundancy > 0:
+                busiest_policy = max(policy_quota, key=lambda name: policy_quota[name])
+                policy_quota[busiest_policy] += redundancy
+
+            for name, quota in policy_quota.items():
+                self._logger.info(
+                    f"policy {name} payload: {self._num_experiences_by_policy[name]},  quota: {quota} node(s)")
+                for i in range(quota):
+                    trainer_id = (i + offset) % self.num_trainers
+                    self._policy2trainer[name].append(f"TRAINER.{trainer_id}")
+                    self._trainer2policies[f"TRAINER.{trainer_id}"].append(name)
+                offset = (offset + quota) % self.num_trainers
+
+        # re-allocation
+        self._logger.info("Re-allocating policy states on trainers...")
+        for trainer_name, policy_names in self._trainer2policies.items():
+            self._proxy.send(
+                SessionMessage(
+                    MsgTag.INIT_POLICY_STATE, self._proxy.name, trainer_name,
+                    body={MsgKey.POLICY_STATE: {name: self.policy_dict[name].get_state() for name in policy_names}}
+                )
+            )
+
+    def update(self, exp_by_policy: Dict[str, ExperienceSet]):
+        exp_to_send, updated = {}, set()
+        for policy_name, exp in exp_by_policy.items():
+            self._num_experiences_by_policy[policy_name] += exp.size
+            self._exp_cache[policy_name].extend(exp)
+            if (
+                self._exp_cache[policy_name].size >= self.update_trigger[policy_name] and
+                self._num_experiences_by_policy[policy_name] >= self.warmup[policy_name]
+            ):
+                exp_to_send[policy_name] = self._exp_cache.pop(policy_name)
+                updated.add(policy_name)
+
```

Comment:
My personal opinion is that it is better to call `allocate_trainers()` here so that the allocation will always base on the experience amount.

Which also means that maybe there is no need to call this function in `__init__` and the end of this function